flocker/control/test/test_httpapi.py

```diff
@@ -45,25 +71,263 @@ def assertGoodResult(self, method, path, expected_good_result):
 
         :return Deferred: Fires when test is done.
         """
-        requesting = self.agent.request(method, path)
+        requesting = self.assertResponseCode(method, path, request_body, OK)
         requesting.addCallback(readBody)
         requesting.addCallback(lambda body: self.assertEqual(
             goodResult(expected_good_result), loads(body)))
         return requesting
 
+    def assertBadResult(self, method, path, request_body,
+                        expected_code, expected_bad_result):
+        """
+        Assert a particular JSON response for the given API request.
+
+        :param bytes method: HTTP method to request.
+        :param bytes path: HTTP path.
+        :param unicode expected_bad_result: The result expected in the error
+            respones.
+
+        :return Deferred: Fires when test is done.
+        """
+        requesting = self.assertResponseCode(
+            method, path, request_body, expected_code)
+        requesting.addCallback(readBody)
+        requesting.addCallback(lambda body: self.assertEqual(
+            badResult(expected_bad_result), loads(body)))
+        return requesting
+
+
+class VersionTestsMixin(APITestsMixin):
+    """
+    Tests for the service version description endpoint at ``/version``.
+    """
     def test_version(self):
         """
         The ``/version`` command returns JSON-encoded ``__version__``.
         """
-        return self.assertGoodResult(b"GET", b"/version",
-                                     {u'flocker': __version__})
+        return self.assertGoodResult(
+            b"GET", b"/version", None, {u'flocker': __version__}
+        )
 
 
 def _build_app(test):
     test.initialize()
     return DatasetAPIUserV1(test.persistence_service).app
-RealTestsAPI, MemoryTestsAPI = buildIntegrationTests(
-    APITestsMixin, "API", _build_app)
+
+RealTestsVersion, MemoryTestsVersion = buildIntegrationTests(
+    VersionTestsMixin, "Version", _build_app)
+
+
+class CreateDatasetTestsMixin(APITestsMixin):
+    """
+    Tests for the dataset creation endpoint at ``/datasets``.
+    """
+    NODE_A = b"10.0.0.1"
+
+    def test_wrong_schema(self):
+        """
+        If a ``POST`` request made to the endpoint includes a body which
+        doesn't match the ``definitions/datasets`` schema, the response is an
+        error indication a validation failure.
+        """
+        return self.assertBadResult(
+            b"POST", b"/datasets", {u"primary": self.NODE_A, u"junk": u"garbage"},
+            BAD_REQUEST,
+            {u'description': u"The provided JSON doesn't match the required schema.",
+             u'errors': [
+                 u"Additional properties are not allowed (u'junk' was unexpected)"]}
+        )
+
+    def test_dataset_id_collision(self):
```

Comment:
:exclamation: There should perhaps be a test where the collision occurs with an unrelated node (i.e. not the one that is going to be hosting the new volume.